pkg/catalogv2/oci/capturewait/client.go

```diff
@@ -0,0 +1,93 @@
+package capturewait
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	RateLimitRemainingHeader = "RateLimit-Remaining"
+)
+
+// Transport is an HTTP transport with capturing duration to wait when 429 is returned by the server.
+type Transport struct {
+	// Base is the underlying HTTP transport to use.
+	// If nil, http.DefaultTransport is used for round trips.
+	Base http.RoundTripper
+
+	// BackOffDuration is used as a duration to wait for by the
+	// OCI handler until the next retry to make a call to the OCI registry.
+	BackOffDuration float64
+
+	// Client is used to make a HEAD request to the registry to get the duration to wait.
+	Client *http.Client
+}
+
+// NewTransport creates an HTTP Transport with the ability to store duration to wait when 429 is returned by the server.
+func NewTransport(base http.RoundTripper, client http.Client) *Transport {
+	return &Transport{
+		Base:   base,
+		Client: &client,
+	}
+}
+
+// RoundTrip executes a single HTTP transaction, returning a Response for the
+// provided Request and it captures the duration to wait until the server unblocks
+// the user to make more requests. It checks the headers to capture the duration.
+func (t *Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	resp, respErr := t.Base.RoundTrip(req)
+	// If we hit 429 status code, try to capture the duration using available headers.
+	if resp != nil && resp.StatusCode == http.StatusTooManyRequests {
+		// Case 1: Check if RateLimit-Remaining is present and get values from it.
+		headerValue := resp.Header.Get(RateLimitRemainingHeader)
+		if headerValue != "" {
+			timeWindow, err := decodeTimeFromHeaderValue(headerValue, " w=")
+			if err != nil {
+				logrus.Error("error:", err)
+				return resp, err
+			}
+			t.BackOffDuration = timeWindow
+		} else {
+			// Case 2: Dockerhub: It sends duration information when a HEAD call is made to the manifest.
```

Comment:
Are we going to always do a HEAD request in case the registry returns 429 but doesn't have the header then?